controllers/namespacelabel_controller.go

```diff
@@ -65,81 +64,101 @@ type NamespaceLabelReconciler struct {
 		if k8serrors.IsNotFound(err) {
 			return ctrl.Result{}, nil
 		}
-		log.WithError(errors.New(err.Error())).WithFields(logrus.Fields{"NamespaceLabel": req.NamespacedName}).Error("unable to fetch NamespaceLabel")
+		log.Error(err, "Unable to fetch NamespaceLabel", "namespaceLabel", req.NamespacedName.String())
 		return ctrl.Result{Requeue: true}, client.IgnoreNotFound(err)
 	}
 
 	var namespace corev1.Namespace
 	if err := r.Client.Get(ctx, types.NamespacedName{Name: namespaceLabel.ObjectMeta.Namespace}, &namespace); err != nil {
-		log.WithError(errors.New(err.Error())).WithFields(logrus.Fields{"namespace": namespaceLabel.ObjectMeta.Namespace}).Error("unable to fetch namespace while getting previous labels")
+		log.Error(err, "Unable to fetch namespace", "namespace", namespaceLabel.ObjectMeta.Namespace)
 		return ctrl.Result{Requeue: true}, client.IgnoreNotFound(err)
 	}
+	if namespaceLabel.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(&namespaceLabel, FinalizerName) {
+			if err := r.addFinalizer(ctx, log, &namespaceLabel, &namespace); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(&namespaceLabel, FinalizerName) {
+			if res, err := r.deleteLabels(ctx, log, &namespaceLabel, &namespace); err != nil {
+				return res, err
+			}
+			return ctrl.Result{}, nil
+		}
+	}
+	if err := r.updateLabels(ctx, log, &namespaceLabel, &namespace); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
 
+func (r *NamespaceLabelReconciler) updateLabels(ctx context.Context, log logr.Logger, namespaceLabel *danaiodanaiov1alpha1.NamespaceLabel, namespace *corev1.Namespace) error {
 	labels := namespace.GetLabels()
-	protectedLabels := getProtectedLabels()
+	protectedPrefixes := getProtectedPrefixes(r.ProtectedPrefixes)
 	for key, val := range namespaceLabel.Spec.Labels {
-		if !utils.Contains(protectedLabels, key) {
+		_, exists := labels[key]
+		if !utils.Contains(protectedPrefixes, strings.Split(key, "/")[0]) && !exists {
 			labels[key] = val
+			condition := metav1.Condition{Type: "LabelApplied", Status: "True", Reason: "LabelApplied", Message: "Label " + key + " = " + val + " was applied", LastTransitionTime: metav1.Time{Time: time.Now()}}
+			namespaceLabel.Status.Conditions = append(namespaceLabel.Status.Conditions, condition)
+			if !utils.Contains(namespaceLabel.Status.EnforcedLabels, key) {
+				namespaceLabel.Status.EnforcedLabels = append(namespaceLabel.Status.EnforcedLabels, key)
+			}
+		} else if exists && !utils.Contains(namespaceLabel.Status.EnforcedLabels, key) {
+			condition := metav1.Condition{Type: "LabelApplied", Status: "False", Reason: "LabelNotApplied", Message: "Label was not applied because a label with the same name was applied earlier", LastTransitionTime: metav1.Time{Time: time.Now()}}
+			namespaceLabel.Status.Conditions = append(namespaceLabel.Status.Conditions, condition)
+			delete(namespaceLabel.Spec.Labels, key)
```

Comment:
I wouldn't delete the label from the `namespaceLabel` object since it would be very confusing for the end-user. I think it'd be better to keep the label and update the status as you already do, which is nice.

Reply:
updated in latest pr